### DISCUSSION TITLE
Proposal: Break `<pf-icon>`'s javascript APIs in 4.0 for better webpack compatibility

### DISCUSSION THREAD

**Discussion by bennypowers** [patternfly/patternfly-elements#2765] [Ideas]
tl;dr:  webpack has trouble with pf-icon and i think the only solution is a breaking change

`PfIcon` contains a dynamic import which references an expression `spec`:

https://github.com/patternfly/patternfly-elements/blob/970227844c2d5c012e78dae71d9d9e9a4ce616af/elements/pf-icon/BaseIcon.ts#L139-L162

in particular line 151 there:

https://github.com/patternfly/patternfly-elements/blob/970227844c2d5c012e78dae71d9d9e9a4ce616af/elements/pf-icon/BaseIcon.ts#L151

This works fine in every runtime, but some bundlers (cough _webpack_) [simply refuse to allow it](https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import).

We can fix this with a breaking change to pf-icon. I don't believe it's possible to do a "soft-deprecation" (i.e. to retain the existing apis but warn when they are used) because the existence of this import expression in the code in any form will cause this problem. the `/* webpackIgnore */` comment does not help.

I've prepared some minimal reproductions (attached) which demonstrate that vite and rollup work fine, but webpack is unfortunately obstinate.

## Vite  👍 Works

<details><summary>Reproduction</summary>

```sh
npm init vite -- --template vanilla vite-project
cd vite-project
npm i @patternfly/elements@3.0.1
echo "import '@patternfly/elements/pf-icon/pf-icon.js';" >> main.js
sed -i '4i <script type="importmap">{"imports":{"@patternfly/icons/":"/node_modules/@patternfly/icons/"}}</script>' index.html
sed -i '11i <pf-icon icon="ad"></pf-icon>' index.html
npm run dev
```

</details>

## Rollup 👍 Works

<details><summary>Reproduction</summary>

```
git clone https://github.com/rollup/rollup-starter-app
cd rollup-starter-app
npm i @patternfly/elements@3.0.1
echo "import '@patternfly/elements/pf-icon/pf-icon.js';" >> src/main.js
sed -i '4i <script type="importmap">{"imports":{"@patternfly/icons/":"/@patternfly/icons/"}}</script>' public/index.html
sed -i '19i <pf-icon icon="ad"></pf-icon>' public/index.html
mkdir -p public/@patternfly
cp -r node_modules/@patternfly/icons public/@patternfly/
npm run build
npm start
```

</details>

## Webpack 💣 Fails so hard

> WARNING in ./node_modules/@patternfly/elements/pf-icon/BaseIcon.js 91:34-46
> Critical dependency: the request of a dependency is an expression

<details><summary>Reproduction</summary>

```
npm init -y
npm i webpack webpack-cli webpack-dev-server @patternfly/elements@3
mkdir dist

echo '<!DOCTYPE html>
<html>
    <head>
        <title>Empty project</title>
        <meta charset="utf-8">
        <script type="importmap">{"imports":{"@patternfly/icons/":"/node_modules/@patternfly/icons/"}}</script>
    </head>
    <body>
        <pf-icon icon="ad"></pf-icon>
        <div id="app"></div>
        <script src="bundle.js"></script>
    </body>
</html>
' > dist/index.html
echo "import '@patternfly/elements/pf-icon/pf-icon.js';" > main.js
echo "const webpack = require('webpack');
const path = require('path');
module.exports = {
  entry: './main.js',
  devServer: {
    static: {
      directory: path.join(__dirname, 'dist'),
    },
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'bundle.js'
  }
};" > webpack.config.js
npx webpack-cli serve --mode development
```

</details>

## Proposal

I propose to remove the overridable static `getIconUrl` method and to change the signature of the static method `addIconSet` in a backwards-incompatible way.
I also propose adding a new overridable static `resolve()` method, similar to the old `getIconUrl`, which would return a promise of the icon node to render, rather than the url to import it from.

**Comment by heyMP** [comment]
@bennypowers I think that's reasonable.  Product Trials mostly relies on RHDC to load and configure PFE icon.  There are a few instances were we reference the icon node directly in our components, but it sounds like the proposed changes won't impact that usage.

Example:
```ts
import * as InfoCircle from '@patternfly/elements/pf-icon/icons/fas/info-circle.js';

  renderTooltip() {
    return html`
      <rh-tooltip part="tooltip">
        ${InfoCircle.default.cloneNode(true)}
        <slot slot="content" name="authenticated-tooltip"></slot>
      </rh-tooltip>
    `;
  }
```

**Reply by bennypowers** [reply]
this snippet is anyways broken since we removed `/icons` in favour of `@patternfly/icons` instead

**Reply by bennypowers** [reply]
one thing i think is worth considering is how we might act in an ssr scenario?

like, we'd ideally just print the icon in the shadow root in that case. lit/ssr support async rendering, but i'd want to be very confident we wouldn't need another breaking change to support it

**Comment by eyevana** [comment]
The catalog uses the `getIconUrl` method so that icons are pulled from the CDN instead. How would we handle after the proposed change? 
```
<script>
    (async function () {
      const PfIcon = await customElements.whenDefined("pf-icon");
      if (PfIcon) PfIcon.getIconUrl = (set, icon) =>
        `https://www.redhatstatic.com/dx/v1-alpha/@patternfly/icons@1.0.2/${set}/${icon}.js`;
    })();
</script>
```

**Reply by bennypowers** [reply]
I'm sure this would need improvement, but what about something like:

```js
// assuming that the new API is called `resolve()`
if ('getIconUrl' in PfIcon) {
  PfIcon.getIconUrl = (set, icon) =>
    `https://www.redhatstatic.com/dx/v1-alpha/@patternfly/icons@1.0.2/${set}/${icon}.js`;
} else if ('resolve' in PfIcon) {
  PfIcon.resolve = async (set, icon) =>
    import(`https://www.redhatstatic.com/dx/v1-alpha/@patternfly/icons@1.0.2/${set}/${icon}.js`)
      .then(mod => mod.default);
}
```

BUT in this particular case, I think the best solution would just be to 

```html
<script type="importmap">
{ "imports": {
  "@patternfly/icons/": "https://www.redhatstatic.com/dx/v1-alpha/@patternfly/icons@1.0.2/"
} }
</script>
```

Which would work with both versions

**Comment by bennypowers** [comment]
@zhawkins 
<img width="314" alt="Screenshot 2024-06-03 at 17 25 53" src="https://github.com/patternfly/patternfly-elements/assets/1466420/9a2a39ad-ca56-4fed-bb4b-82c294e14318">

thanks!

**Reply by zhawkins** [reply]
yup, sorry. This is fine with us too 👍
